README

// only few comments, yet ;)

Driver needs companion userspace applications from http://code.google.com/p/dreamtuner/

Devices permissions:

  Driver is registering the following devices:

    DVB:

    /dev/dvb/adapterX/demux0
    /dev/dvb/adapterX/dvr0
    /dev/dvb/adapterX/frontend0 [registered later]

    controlling:

    /dev/vtunerX

    where X is ordered by driver installation.

  All devices get default device permissions, what usually
  can't be exactly what we need (like 660 root/root).

  There is no simple way of changing permissions of
  created files from driver. 

  The only right(tm) way is using some upper layer
  subsystem. What means, int time of writing such notes,
  writing udev rules.

  This rule works:

  KERNEL=="vtunerc*", MODE="0666"
  
  put it in /etc/udev/rules.d/10-local.rules (in Gentoo)

  ...
  ...
  ...
  
  How to write udev rule: http://hackaday.com/2009/09/18/how-to-write-udev-rules/

  Bug on Fedora: http://www.linuxquestions.org/questions/fedora-35/permissions-on-dvb-card-432770/


HISTORY
r0: 2011-06-23
Initial version publiced on linux-media on 19th July 2011 (slightly modified by removing backward compatibility)

r2: 2011-06-28
Added DVB-C support. Thanks to Gerald

r3: 2011-06-30
v1.1 - write() is now reading only whole TS packets. Added module parameter 'tscheck=1' to enable TS data checking


vtuner.h
/*
 * vtunerc: /dev/vtunerc API
 *
 * Copyright (C) 2010-11 Honza Petrous <jpetrous@smartimp.cz>
 * [based on dreamtuner userland code by Roland Mieslinger]
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation version 2.
 *
 * This program is distributed WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#ifndef _VTUNER_H_
#define _VTUNER_H_

#include <linux/dvb/version.h>
#include <linux/dvb/frontend.h>
#include <linux/dvb/dmx.h>

#define VT_S   0x01
#define VT_C   0x02
#define VT_T   0x04
#define VT_S2  0x08

#define MSG_SET_FRONTEND		1
#define MSG_GET_FRONTEND		2
#define MSG_READ_STATUS			3
#define MSG_READ_BER			4
#define MSG_READ_SIGNAL_STRENGTH	5
#define MSG_READ_SNR			6
#define MSG_READ_UCBLOCKS		7
#define MSG_SET_TONE			8
#define MSG_SET_VOLTAGE			9
#define MSG_ENABLE_HIGH_VOLTAGE		10
#define MSG_SEND_DISEQC_MSG		11
#define MSG_SEND_DISEQC_BURST		13
#define MSG_PIDLIST			14
#define MSG_TYPE_CHANGED		15
#define MSG_SET_PROPERTY		16
#define MSG_GET_PROPERTY		17

#define MSG_NULL			1024
#define MSG_DISCOVER			1025
#define MSG_UPDATE       		1026

struct diseqc_master_cmd {
	u8 msg[6];
	u8 msg_len;
};

struct vtuner_message {
	s32 type;
	union {
		struct {
			u32	frequency;
			u8	inversion;
			union {
				struct {
					u32	symbol_rate;
					u32	fec_inner;
				} qpsk;
				struct {
					u32   symbol_rate;
					u32   fec_inner;
					u32	modulation;
				} qam;
				struct {
					u32	bandwidth;
					u32	code_rate_HP;
					u32	code_rate_LP;
					u32	constellation;
					u32	transmission_mode;
					u32	guard_interval;
					u32	hierarchy_information;
				} ofdm;
				struct {
					u32	modulation;
				} vsb;
			} u;
		} fe_params;
		struct dtv_property prop;
		u32 status;
		u32 ber;
		u16 ss;
		u16 snr;
		u32 ucb;
		u8 tone;
		u8 voltage;
		struct diseqc_master_cmd diseqc_master_cmd;
		u8 burst;
		u16 pidlist[30];
		u8  pad[72];
		u32 type_changed;
	} body;
};

#define VTUNER_MAJOR		226

/*#define PVR_FLUSH_BUFFER	_IO(VTUNER_MAJOR, 0)*/
#define VTUNER_GET_MESSAGE	_IOR(VTUNER_MAJOR, 1, struct vtuner_message *)
#define VTUNER_SET_RESPONSE 	_IOW(VTUNER_MAJOR, 2, struct vtuner_message *)
#define VTUNER_SET_NAME		_IOW(VTUNER_MAJOR, 3, char *)
#define VTUNER_SET_TYPE		_IOW(VTUNER_MAJOR, 4, char *)
#define VTUNER_SET_FE_INFO	_IOW(VTUNER_MAJOR, 6, struct dvb_frontend_info *)
#define VTUNER_SET_NUM_MODES	_IOW(VTUNER_MAJOR, 7, int)
#define VTUNER_SET_MODES	_IOW(VTUNER_MAJOR, 8, char *)

#endif



vtunerc_main.c
/*
 * vtunerc: Virtual adapter driver
 *
 * Copyright (C) 2010-11 Honza Petrous <jpetrous@smartimp.cz>
 * [Created 2010-03-23]
 * Sponsored by Smartimp s.r.o. for its NessieDVB.com box
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation version 2.
 *
 * This program is distributed WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <linux/module.h>	/* Specifically, a module */
#include <linux/kernel.h>	/* We're doing kernel work */
#include <linux/proc_fs.h>
#include <linux/init.h>
#include <linux/i2c.h>
#include <asm/uaccess.h>
#include <linux/delay.h>

#include "demux.h"
#include "dmxdev.h"
#include "dvb_demux.h"
#include "dvb_frontend.h"
#include "dvb_net.h"
#include "dvbdev.h"

#include "vtunerc_priv.h"

#define VTUNERC_MODULE_VERSION "1.1p3"

DVB_DEFINE_MOD_OPT_ADAPTER_NR(adapter_nr);

#define DRIVER_NAME		"vTuner proxy"

#define VTUNERC_PROC_FILENAME	"vtunerc%i"

#define VTUNERC_MAX_ADAPTERS	4

static struct vtunerc_ctx *vtunerc_tbl[VTUNERC_MAX_ADAPTERS] = { NULL };

/* module params */
static struct vtunerc_config config = {
	.devices = 1,
	.tscheck = 0,
	.debug = 0
};

static int pidtab_find_index(unsigned short *pidtab, int pid)
{
	int i = 0;

	while (i < MAX_PIDTAB_LEN) {
		if (pidtab[i] == pid)
			return i;
		i++;
	}

	return -1;
}

static int pidtab_add_pid(unsigned short *pidtab, int pid)
{
	int i;

	/* TODO: speed-up hint: add pid sorted */

	for (i = 0; i < MAX_PIDTAB_LEN; i++)
		if (pidtab[i] == PID_UNKNOWN) {
			pidtab[i] = pid;
			return 0;
		}

	return -1;
}

static int pidtab_del_pid(unsigned short *pidtab, int pid)
{
	int i;

	/* TODO: speed-up hint: delete sorted */

	for (i = 0; i < MAX_PIDTAB_LEN; i++)
		if (pidtab[i] == pid) {
			pidtab[i] = PID_UNKNOWN;
			/* TODO: move rest */
			return 0;
		}

	return -1;
}

static void pidtab_copy_to_msg(struct vtunerc_ctx *ctx,
				struct vtuner_message *msg)
{
	int i;

	for (i = 0; i < (MAX_PIDTAB_LEN - 1); i++)
		msg->body.pidlist[i] = ctx->pidtab[i]; /*TODO: optimize it*/
	msg->body.pidlist[MAX_PIDTAB_LEN - 1] = 0;
}

static int vtunerc_start_feed(struct dvb_demux_feed *feed)
{
	struct dvb_demux *demux = feed->demux;
	struct vtunerc_ctx *ctx = demux->priv;
	struct vtuner_message msg;

	switch (feed->type) {
	case DMX_TYPE_TS:
		break;
	case DMX_TYPE_SEC:
		break;
	case DMX_TYPE_PES:
		printk(KERN_ERR "vtunerc%d: feed type PES is not supported\n",
				ctx->idx);
		return -EINVAL;
	default:
		printk(KERN_ERR "vtunerc%d: feed type %d is not supported\n",
				ctx->idx, feed->type);
		return -EINVAL;
	}

	/* organize PID list table */

	if (pidtab_find_index(ctx->pidtab, feed->pid) < 0) {
		pidtab_add_pid(ctx->pidtab, feed->pid);

		pidtab_copy_to_msg(ctx, &msg);

		msg.type = MSG_PIDLIST;
		vtunerc_ctrldev_xchange_message(ctx, &msg, 0);
	}

	return 0;
}

static int vtunerc_stop_feed(struct dvb_demux_feed *feed)
{
	struct dvb_demux *demux = feed->demux;
	struct vtunerc_ctx *ctx = demux->priv;
	struct vtuner_message msg;

	/* organize PID list table */

	if (pidtab_find_index(ctx->pidtab, feed->pid) > -1) {
		pidtab_del_pid(ctx->pidtab, feed->pid);

		pidtab_copy_to_msg(ctx, &msg);

		msg.type = MSG_PIDLIST;
		vtunerc_ctrldev_xchange_message(ctx, &msg, 0);
	}

	return 0;
}

/* ----------------------------------------------------------- */


#ifdef CONFIG_PROC_FS

static char *get_fe_name(struct dvb_frontend_info *feinfo)
{
	return (feinfo && feinfo->name) ? feinfo->name : "(not set)";
}

/**
 * @brief  procfs file handler
 * @param  buffer:
 * @param  start:
 * @param  offset:
 * @param  size:
 * @param  eof:
 * @param  data:
 * @return =0: success <br/>
 *         <0: if any error occur
 */
#define MAXBUF 512
int vtunerc_read_proc(char *buffer, char **start, off_t offset, int size,
			int *eof, void *data)
{
	char outbuf[MAXBUF] = "[ vtunerc driver, version "
				VTUNERC_MODULE_VERSION " ]\n";
	int blen, i, pcnt;
	struct vtunerc_ctx *ctx = (struct vtunerc_ctx *)data;

	blen = strlen(outbuf);
	sprintf(outbuf+blen, "  sessions: %u\n", ctx->stat_ctrl_sess);
	blen = strlen(outbuf);
	sprintf(outbuf+blen, "  TS data : %u\n", ctx->stat_wr_data);
	blen = strlen(outbuf);
	sprintf(outbuf+blen, "  PID tab :");
	pcnt = 0;
	for (i = 0; i < MAX_PIDTAB_LEN; i++) {
		blen = strlen(outbuf);
		if (ctx->pidtab[i] != PID_UNKNOWN) {
			sprintf(outbuf+blen, " %x", ctx->pidtab[i]);
			pcnt++;
		}
	}
	blen = strlen(outbuf);
	sprintf(outbuf+blen, " (len=%d)\n", pcnt);
	blen = strlen(outbuf);
	sprintf(outbuf+blen, "  FE type : %s\n", get_fe_name(ctx->feinfo));

	blen = strlen(outbuf);

	if (size < blen)
		return -EINVAL;

	if (offset != 0)
		return 0;

	strcpy(buffer, outbuf);

	/* signal EOF */
	*eof = 1;

	return blen;

}
#endif

static char *my_strdup(const char *s)
{
	char *rv = kmalloc(strlen(s)+1, GFP_KERNEL);
	if (rv)
		strcpy(rv, s);
	return rv;
}

struct vtunerc_ctx *vtunerc_get_ctx(int minor)
{
	if (minor >= VTUNERC_MAX_ADAPTERS)
		return NULL;

	return vtunerc_tbl[minor];
}

static int __init vtunerc_init(void)
{
	struct vtunerc_ctx *ctx = NULL;
	struct dvb_demux *dvbdemux;
	struct dmx_demux *dmx;
	int ret = -EINVAL, i, idx;

	printk(KERN_INFO "virtual DVB adapter driver, version "
			VTUNERC_MODULE_VERSION
			", (c) 2010-11 Honza Petrous, SmartImp.cz\n");

	request_module("dvb-core"); /* FIXME: dunno which way it should work :-/ */

	for (idx = 0; idx < config.devices; idx++) {
		ctx = kzalloc(sizeof(struct vtunerc_ctx), GFP_KERNEL);
		if (!ctx)
			return -ENOMEM;

		vtunerc_tbl[idx] = ctx;

		ctx->idx = idx;
		ctx->config = &config;
		ctx->ctrldev_request.type = -1;
		ctx->ctrldev_response.type = -1;
		init_waitqueue_head(&ctx->ctrldev_wait_request_wq);
		init_waitqueue_head(&ctx->ctrldev_wait_response_wq);

		/* dvb */

		/* create new adapter */
		ret = dvb_register_adapter(&ctx->dvb_adapter, DRIVER_NAME,
					   THIS_MODULE, NULL, adapter_nr);
		if (ret < 0)
			goto err_kfree;

		ctx->dvb_adapter.priv = ctx;

		memset(&ctx->demux, 0, sizeof(ctx->demux));
		dvbdemux = &ctx->demux;
		dvbdemux->priv = ctx;
		dvbdemux->filternum = MAX_PIDTAB_LEN;
		dvbdemux->feednum = MAX_PIDTAB_LEN;
		dvbdemux->start_feed = vtunerc_start_feed;
		dvbdemux->stop_feed = vtunerc_stop_feed;
		dvbdemux->dmx.capabilities = 0;
		ret = dvb_dmx_init(dvbdemux);
		if (ret < 0)
			goto err_dvb_unregister_adapter;

		dmx = &dvbdemux->dmx;

		ctx->hw_frontend.source = DMX_FRONTEND_0;
		ctx->mem_frontend.source = DMX_MEMORY_FE;
		ctx->dmxdev.filternum = MAX_PIDTAB_LEN;
		ctx->dmxdev.demux = dmx;

		ret = dvb_dmxdev_init(&ctx->dmxdev, &ctx->dvb_adapter);
		if (ret < 0)
			goto err_dvb_dmx_release;

		ret = dmx->add_frontend(dmx, &ctx->hw_frontend);
		if (ret < 0)
			goto err_dvb_dmxdev_release;

		ret = dmx->add_frontend(dmx, &ctx->mem_frontend);
		if (ret < 0)
			goto err_remove_hw_frontend;

		ret = dmx->connect_frontend(dmx, &ctx->hw_frontend);
		if (ret < 0)
			goto err_remove_mem_frontend;

		sema_init(&ctx->xchange_sem, 1);
		sema_init(&ctx->ioctl_sem, 1);
		sema_init(&ctx->tswrite_sem, 1);

		/* init pid table */
		for (i = 0; i < MAX_PIDTAB_LEN; i++)
			ctx->pidtab[i] = PID_UNKNOWN;

#ifdef CONFIG_PROC_FS
		{
			char procfilename[64];

			sprintf(procfilename, VTUNERC_PROC_FILENAME,
					ctx->idx);
			ctx->procname = my_strdup(procfilename);
			if (create_proc_read_entry(ctx->procname, 0, NULL,
							vtunerc_read_proc,
							ctx) == 0)
				printk(KERN_WARNING
					"vtunerc%d: Unable to register '%s' proc file\n",
					ctx->idx, ctx->procname);
		}
#endif
	}

	vtunerc_register_ctrldev(ctx);

out:
	return ret;

	dmx->disconnect_frontend(dmx);
err_remove_mem_frontend:
	dmx->remove_frontend(dmx, &ctx->mem_frontend);
err_remove_hw_frontend:
	dmx->remove_frontend(dmx, &ctx->hw_frontend);
err_dvb_dmxdev_release:
	dvb_dmxdev_release(&ctx->dmxdev);
err_dvb_dmx_release:
	dvb_dmx_release(dvbdemux);
err_dvb_unregister_adapter:
	dvb_unregister_adapter(&ctx->dvb_adapter);
err_kfree:
	kfree(ctx);
	goto out;
}

static void __exit vtunerc_exit(void)
{
	struct dvb_demux *dvbdemux;
	struct dmx_demux *dmx;
	int idx;

	vtunerc_unregister_ctrldev(&config);

	for (idx = 0; idx < config.devices; idx++) {
		struct vtunerc_ctx *ctx = vtunerc_tbl[idx];
		if(!ctx)
			continue;
		vtunerc_tbl[idx] = NULL;
#ifdef CONFIG_PROC_FS
		remove_proc_entry(ctx->procname, NULL);
		kfree(ctx->procname);
#endif

		vtunerc_frontend_clear(ctx);

		dvbdemux = &ctx->demux;
		dmx = &dvbdemux->dmx;

		dmx->disconnect_frontend(dmx);
		dmx->remove_frontend(dmx, &ctx->mem_frontend);
		dmx->remove_frontend(dmx, &ctx->hw_frontend);
		dvb_dmxdev_release(&ctx->dmxdev);
		dvb_dmx_release(dvbdemux);
		dvb_unregister_adapter(&ctx->dvb_adapter);
		kfree(ctx);
	}

	printk(KERN_NOTICE "vtunerc: unloaded successfully\n");
}

module_init(vtunerc_init);
module_exit(vtunerc_exit);

MODULE_AUTHOR("Honza Petrous");
MODULE_DESCRIPTION("virtual DVB device");
MODULE_LICENSE("GPL");
MODULE_VERSION(VTUNERC_MODULE_VERSION);

module_param_named(devices, config.devices, int, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
MODULE_PARM_DESC(devices, "Number of virtual adapters (default is 1)");

module_param_named(tscheck, config.tscheck, int, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
MODULE_PARM_DESC(tscheck, "Check TS packet validity (default is 0)");

module_param_named(debug, config.debug, int, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
MODULE_PARM_DESC(debug, "Enable debug messages (default is 0)");



vtunerc_priv.h
/*
 * vtunerc: Internal defines
 *
 * Copyright (C) 2010-11 Honza Petrous <jpetrous@smartimp.cz>
 * [Created 2010-03-23]
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation version 2.
 *
 * This program is distributed WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#ifndef _VTUNERC_PRIV_H
#define _VTUNERC_PRIV_H

#include <linux/module.h>	/* Specifically, a module */
#include <linux/kernel.h>	/* We're doing kernel work */
#include <linux/cdev.h>

#include "demux.h"
#include "dmxdev.h"
#include "dvb_demux.h"
#include "dvb_frontend.h"
#include "dvb_net.h"
#include "dvbdev.h"

#include "vtuner.h"

#define MAX_PIDTAB_LEN 30

#define PID_UNKNOWN 0x0FFFF

#define MAX_NUM_VTUNER_MODES 3

struct vtunerc_config {

	int debug;
	int tscheck;
	int devices;
};

struct vtunerc_ctx {

	/* DVB api */
	struct dmx_frontend hw_frontend;
	struct dmx_frontend mem_frontend;
	struct dmxdev dmxdev;
	struct dmxdev dmxdev1;
	struct dmxdev dmxdev2;
	struct dvb_adapter dvb_adapter;
	struct dvb_demux demux;
	struct dvb_frontend *fe;
	struct dvb_net dvbnet;
	struct dvb_device *ca;

	/* internals */
	int idx;
	char *name;
	u8 vtype;
	struct dvb_frontend_info *feinfo;
	struct vtunerc_config *config;

	unsigned short pidtab[MAX_PIDTAB_LEN];

	struct semaphore xchange_sem;
	struct semaphore ioctl_sem;
	struct semaphore tswrite_sem;
	int fd_opened;
	int closing;

	char *procname;

	/* ctrldev */
	char trail[188];
	unsigned int trailsize;
	int noresponse;
	int num_modes;
	char *ctypes[MAX_NUM_VTUNER_MODES];
	struct vtuner_message ctrldev_request;
	struct vtuner_message ctrldev_response;
	wait_queue_head_t ctrldev_wait_request_wq;
	wait_queue_head_t ctrldev_wait_response_wq;

	/* proc statistics */
	unsigned int stat_wr_data;
	unsigned int stat_rd_data;
	unsigned int stat_ctrl_sess;
	unsigned short pidstat[MAX_PIDTAB_LEN];
};

int vtunerc_register_ctrldev(struct vtunerc_ctx *ctx);
void vtunerc_unregister_ctrldev(struct vtunerc_config *config);
struct vtunerc_ctx *vtunerc_get_ctx(int minor);
int /*__devinit*/ vtunerc_frontend_init(struct vtunerc_ctx *ctx);
int /*__devinit*/ vtunerc_frontend_clear(struct vtunerc_ctx *ctx);
int vtunerc_ctrldev_xchange_message(struct vtunerc_ctx *ctx,
					struct vtuner_message *msg,
					int wait4response);
#define dprintk(ctx, fmt, arg...) do {					\
if (ctx->config && (*ctx->config.debug))				\
	printk(KERN_DEBUG "vtunerc%d: %s", ctx->idx, fmt, ## arg);	\
} while (0)

#endif


vtunerc_ctrldev.c
/*
 * vtunerc: /dev/vtunerc device
 *
 * Copyright (C) 2010-11 Honza Petrous <jpetrous@smartimp.cz>
 * [Created 2010-03-23]
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation version 2.
 *
 * This program is distributed WITHOUT ANY WARRANTY of any
 * kind, whether express or implied; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <linux/kernel.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/fs.h>
#include <linux/delay.h>

#include <linux/time.h>
#include <linux/poll.h>

#include "vtunerc_priv.h"

#define VTUNERC_CTRLDEV_MAJOR	266
#define VTUNERC_CTRLDEV_NAME	"vtunerc"

#define VTUNER_MSG_LEN (sizeof(struct vtuner_message))

static ssize_t vtunerc_ctrldev_write(struct file *filp, const char *buff,
					size_t len, loff_t *off)
{
	struct vtunerc_ctx *ctx = filp->private_data;
	struct dvb_demux *demux = &ctx->demux;
	char *kernel_buf;
	int tailsize = len % 188;

	if (ctx->closing)
		return -EINTR;

	if (len < 188) {
		printk(KERN_ERR "vtunerc%d: Data are shorter then TS packet size (188B)\n",
				ctx->idx);
		return -EINVAL;
	}

	len -= tailsize;
	kernel_buf = kmalloc(len, GFP_KERNEL);

	if (kernel_buf == NULL)
		return -ENOMEM;

	if (down_interruptible(&ctx->tswrite_sem))
		return -ERESTARTSYS;

	if (copy_from_user(kernel_buf, buff, len)) {
		printk(KERN_ERR "vtunerc%d: userdata passing error\n",
				ctx->idx);
		up(&ctx->tswrite_sem);
		return -EINVAL;
	}

	if (ctx->config->tscheck) {
		int i;

		for (i = 0; i < len; i += 188)
			if (kernel_buf[i] != 0x47) { /* start of TS packet */
				printk(KERN_ERR "vtunerc%d: Data not start on packet boundary: index=%d data=%02x %02x %02x %02x %02x ...\n",
						ctx->idx, i / 188, kernel_buf[i], kernel_buf[i + 1],
						kernel_buf[i + 2], kernel_buf[i + 3], kernel_buf[i + 4]);
				up(&ctx->tswrite_sem);
				return -EINVAL;
			}
	}

	ctx->stat_wr_data += len;
	dvb_dmx_swfilter_packets(demux, kernel_buf, len / 188);

	up(&ctx->tswrite_sem);

#ifdef CONFIG_PROC_FS
	/* TODO:  analyze injected data for statistics */
#endif

	kfree(kernel_buf);

	return len;
}

static ssize_t vtunerc_ctrldev_read(struct file *filp, char __user *buff,
		size_t len, loff_t *off)
{
	struct vtunerc_ctx *ctx = filp->private_data;

	ctx->stat_rd_data += len;

	/* read op is not using in current vtuner protocol */
	return 0 ;
}

static int vtunerc_ctrldev_open(struct inode *inode, struct file *filp)
{
	struct vtunerc_ctx *ctx;
	int minor;

	minor = MINOR(inode->i_rdev);
	ctx = filp->private_data = vtunerc_get_ctx(minor);
	if (ctx == NULL)
		return -ENOMEM;

	ctx->stat_ctrl_sess++;

	/*FIXME: clear pidtab */

	ctx->fd_opened++;
	ctx->closing = 0;

	return 0;
}

static int vtunerc_ctrldev_close(struct inode *inode, struct file *filp)
{
	struct vtunerc_ctx *ctx = filp->private_data;
	int minor;
	struct vtuner_message fakemsg;

	ctx->fd_opened--;
	ctx->closing = 1;

	minor = MINOR(inode->i_rdev);

	/* set FAKE response, to allow finish any waiters
	   in vtunerc_ctrldev_xchange_message() */
	ctx->ctrldev_response.type = 0;
	wake_up_interruptible(&ctx->ctrldev_wait_response_wq);

	/* clear pidtab */
	if (down_interruptible(&ctx->xchange_sem))
		return -ERESTARTSYS;
	memset(&fakemsg, 0, sizeof(fakemsg));
	vtunerc_ctrldev_xchange_message(ctx, &fakemsg, 0);
	up(&ctx->xchange_sem);

	return 0;
}

static long vtunerc_ctrldev_ioctl(struct file *file, unsigned int cmd,
					unsigned long arg)
{
	struct vtunerc_ctx *ctx = file->private_data;
	int len, ret = 0;
	int i;

	if (ctx->closing)
		return -EINTR;

	if (down_interruptible(&ctx->ioctl_sem))
		return -ERESTARTSYS;

	switch (cmd) {
	case VTUNER_SET_NAME:
		len = strlen((char *)arg) + 1;
		ctx->name = kmalloc(len, GFP_KERNEL);
		if (ctx->name == NULL) {
			printk(KERN_ERR "vtunerc%d: no memory\n", ctx->idx);
			ret = -ENOMEM;
			break;
		}
		if (copy_from_user(ctx->name, (char *)arg, len)) {
			ret = -EFAULT;
			break;
		}
		break;

	case VTUNER_SET_MODES:
		for (i = 0; i < ctx->num_modes; i++)
			ctx->ctypes[i] = &(((char *)(arg))[i*32]);
		if (ctx->num_modes != 1) {
			printk(KERN_ERR "vtunerc%d: currently supported only num_modes = 1!\n",
					ctx->idx);
			ret = -EINVAL;
			break;
		}
		/* follow into old code for compatibility */

	case VTUNER_SET_TYPE:
		if (strcasecmp((char *)arg, "DVB-S") == 0) {
			ctx->vtype = VT_S;
			printk(KERN_NOTICE "vtunerc%d: setting DVB-S tuner vtype\n",
					ctx->idx);
		} else
		if (strcasecmp((char *)arg, "DVB-S2") == 0) {
			ctx->vtype = VT_S2;
			printk(KERN_NOTICE "vtunerc%d: setting DVB-S2 tuner vtype\n",
					ctx->idx);
		} else
		if (strcasecmp((char *)arg, "DVB-T") == 0) {
			ctx->vtype = VT_T;
			printk(KERN_NOTICE "vtunerc%d: setting DVB-T tuner vtype\n",
					ctx->idx);
		} else
		if (strcasecmp((char *)arg, "DVB-C") == 0) {
			ctx->vtype = VT_C;
			printk(KERN_NOTICE "vtunerc%d: setting DVB-C tuner vtype\n",
					ctx->idx);
		} else {
			printk(KERN_ERR "vtunerc%d: unregognized tuner vtype '%s'\n",
					ctx->idx, (char *)arg);
			ret = -ENODEV;
			break;
		}

		if ((vtunerc_frontend_init(ctx))) {
			ctx->vtype = 0;
			printk(KERN_ERR "vtunerc%d: failed to initialize tuner's internals\n",
					ctx->idx);
			ret = -ENODEV;
			break;
		}

		break;


	case VTUNER_SET_FE_INFO:
		len = sizeof(struct dvb_frontend_info);
		ctx->feinfo = kmalloc(len, GFP_KERNEL);
		if (ctx->feinfo == NULL) {
			printk(KERN_ERR "vtunerc%d: no mem\n", ctx->idx);
			ret = -ENOMEM;
			break;
		}
		if (copy_from_user(ctx->feinfo, (char *)arg, len)) {
			ret = -EFAULT;
			break;
		}
		break;

	case VTUNER_GET_MESSAGE:
		if (wait_event_interruptible(ctx->ctrldev_wait_request_wq,
					ctx->ctrldev_request.type != -1)) {
			ret = -ERESTARTSYS;
			break;
		}

		BUG_ON(ctx->ctrldev_request.type == -1);

		if (copy_to_user((char *)arg, &ctx->ctrldev_request,
					VTUNER_MSG_LEN)) {
			ret = -EFAULT;
			break;
		}

		ctx->ctrldev_request.type = -1;

		if (ctx->noresponse)
			up(&ctx->xchange_sem);

		break;

	case VTUNER_SET_RESPONSE:
		if (copy_from_user(&ctx->ctrldev_response, (char *)arg,
					VTUNER_MSG_LEN)) {
			ret = -EFAULT;
		}
		wake_up_interruptible(&ctx->ctrldev_wait_response_wq);

		break;

	case VTUNER_SET_NUM_MODES:
		ctx->num_modes = (int) arg;
		break;

	default:
		printk(KERN_ERR "vtunerc%d: unknown IOCTL 0x%x\n", ctx->idx, cmd);
		ret = -ENOTTY; /* Linus: the only correct one return value for unsupported ioctl */

		break;
	}
	up(&ctx->ioctl_sem);

	return ret;
}

static unsigned int vtunerc_ctrldev_poll(struct file *filp, poll_table *wait)
{
	struct vtunerc_ctx *ctx = filp->private_data;
	unsigned int mask = 0;

	if (ctx->closing)
		return -EINTR;

	poll_wait(filp, &ctx->ctrldev_wait_request_wq, wait);

	if (ctx->ctrldev_request.type >= -1 ||
			ctx->ctrldev_response.type >= -1) {
		mask = POLLPRI;
	}

  return mask;
}

/* ------------------------------------------------ */

static const struct file_operations vtunerc_ctrldev_fops = {
	.owner = THIS_MODULE,
	.unlocked_ioctl = vtunerc_ctrldev_ioctl,
	.write = vtunerc_ctrldev_write,
	.read  = vtunerc_ctrldev_read,
	.poll  = (void *) vtunerc_ctrldev_poll,
	.open  = vtunerc_ctrldev_open,
	.release  = vtunerc_ctrldev_close
};

static struct class *pclass;
static struct cdev cdev;
static dev_t chdev;

int vtunerc_register_ctrldev(struct vtunerc_ctx *ctx)
{
	int idx;

	chdev = MKDEV(VTUNERC_CTRLDEV_MAJOR, 0);

	if (register_chrdev_region(chdev, ctx->config->devices, VTUNERC_CTRLDEV_NAME)) {
		printk(KERN_ERR "vtunerc%d: unable to get major %d\n",
				ctx->idx, VTUNERC_CTRLDEV_MAJOR);
		return -EINVAL;
	}

	cdev_init(&cdev, &vtunerc_ctrldev_fops);

	cdev.owner = THIS_MODULE;
	cdev.ops = &vtunerc_ctrldev_fops;

	if (cdev_add(&cdev, chdev, ctx->config->devices) < 0)
		printk(KERN_WARNING "vtunerc%d: unable to create dev\n",
				ctx->idx);

	pclass = class_create(THIS_MODULE, "vtuner");
	if (IS_ERR(pclass)) {
		printk(KERN_ERR "vtunerc%d: unable to register major %d\n",
				ctx->idx, VTUNERC_CTRLDEV_MAJOR);
		return PTR_ERR(pclass);
	}

	for (idx = 0; idx < ctx->config->devices; idx++) {
		struct device *clsdev;

		clsdev = device_create(pclass, NULL,
				MKDEV(VTUNERC_CTRLDEV_MAJOR, idx),
				/*ctx*/ NULL, "vtunerc%d", idx);

		printk(KERN_NOTICE "vtunerc: registered /dev/vtunerc%d\n",
				idx);
	}

	return 0;
}

void vtunerc_unregister_ctrldev(struct vtunerc_config *config)
{
	int idx;

	printk(KERN_NOTICE "vtunerc: unregistering\n");

	unregister_chrdev_region(chdev, config->devices);

	for (idx = 0; idx < config->devices; idx++)
		device_destroy(pclass, MKDEV(VTUNERC_CTRLDEV_MAJOR, idx));

	cdev_del(&cdev);

	class_destroy(pclass);
}


int vtunerc_ctrldev_xchange_message(struct vtunerc_ctx *ctx,
		struct vtuner_message *msg, int wait4response)
{
	if (down_interruptible(&ctx->xchange_sem))
		return -ERESTARTSYS;

	if (ctx->fd_opened < 1) {
		up(&ctx->xchange_sem);
		return 0;
	}

#if 0
	BUG_ON(ctx->ctrldev_request.type != -1);
#else
	if(ctx->ctrldev_request.type != -1)
		printk(KERN_WARNING "orphan request type %d detected\n", ctx->ctrldev_request.type);

#endif

	memcpy(&ctx->ctrldev_request, msg, sizeof(struct vtuner_message));
	ctx->ctrldev_response.type = -1;
	ctx->noresponse = !wait4response;
	wake_up_interruptible(&ctx->ctrldev_wait_request_wq);

	if (!wait4response)
		return 0;

	if (wait_event_interruptible(ctx->ctrldev_wait_response_wq,
				ctx->ctrldev_response.type != -1)) {
		ctx->ctrldev_request.type = -1;
		up(&ctx->xchange_sem);
		return -ERESTARTSYS;
	}

	BUG_ON(ctx->ctrldev_response.type == -1);

	memcpy(msg, &ctx->ctrldev_response, sizeof(struct vtuner_message));
	ctx->ctrldev_response.type = -1;

	up(&ctx->xchange_sem);

	return 0;
}
